ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,2157 @@
+use std::ffi::c_void;
+use std::iter::once;
+
+use deno_core::v8::fast_api;
+use dynasmrt::dynasm;
+use dynasmrt::DynasmApi;
+use dynasmrt::ExecutableBuffer;
+
+use crate::needs_unwrap;
+use crate::NativeType;
+use crate::Symbol;
+
+pub(crate) fn is_compatible(sym: &Symbol) -> bool {
+  cfg!(any(
+    all(target_arch = "x86_64", target_family = "unix"),
+    all(target_arch = "x86_64", target_family = "windows"),
+    all(target_arch = "aarch64", target_vendor = "apple")
+  )) && !sym.can_callback
+}
+
+pub(crate) fn compile_trampoline(sym: &Symbol) -> Trampoline {
+  #[cfg(all(target_arch = "x86_64", target_family = "unix"))]
+  return SysVAmd64::compile(sym);
+  #[cfg(all(target_arch = "x86_64", target_family = "windows"))]
+  return Win64::compile(sym);
+  #[cfg(all(target_arch = "aarch64", target_vendor = "apple"))]
+  return Aarch64Apple::compile(sym);
+  #[allow(unreachable_code)]
+  {
+    unimplemented!("fast API is not implemented for the current target");
+  }
+}
+
+pub(crate) fn make_template(sym: &Symbol, trampoline: &Trampoline) -> Template {
+  let mut params = once(fast_api::Type::V8Value)
+    .chain(sym.parameter_types.iter().map(|t| t.into()))
+    .collect::<Vec<_>>();
+
+  let ret = if needs_unwrap(sym.result_type) {
+    params.push(fast_api::Type::TypedArray(fast_api::CType::Int32));
+    fast_api::Type::Void
+  } else {
+    fast_api::Type::from(&sym.result_type)
+  };
+
+  Template {
+    args: params.into_boxed_slice(),
+    ret: (&ret).into(),
+    symbol_ptr: trampoline.ptr(),
+  }
+}
+
+/// Trampoline for fast-call FFI functions
+///
+/// Calls the FFI function without the first argument (the receiver)
+pub(crate) struct Trampoline(ExecutableBuffer);
+
+impl Trampoline {
+  fn ptr(&self) -> *const c_void {
+    &self.0[0] as *const u8 as *const c_void
+  }
+}
+
+pub(crate) struct Template {
+  args: Box<[fast_api::Type]>,
+  ret: fast_api::CType,
+  symbol_ptr: *const c_void,
+}
+
+impl fast_api::FastFunction for Template {
+  fn function(&self) -> *const c_void {
+    self.symbol_ptr
+  }
+
+  fn args(&self) -> &'static [fast_api::Type] {
+    Box::leak(self.args.clone())
+  }
+
+  fn return_type(&self) -> fast_api::CType {
+    self.ret
+  }
+}
+
+impl From<&NativeType> for fast_api::Type {
+  fn from(native_type: &NativeType) -> Self {
+    match native_type {
+      NativeType::Bool => fast_api::Type::Bool,
+      NativeType::U8 | NativeType::U16 | NativeType::U32 => {
+        fast_api::Type::Uint32
+      }
+      NativeType::I8 | NativeType::I16 | NativeType::I32 => {
+        fast_api::Type::Int32
+      }
+      NativeType::F32 => fast_api::Type::Float32,
+      NativeType::F64 => fast_api::Type::Float64,
+      NativeType::Void => fast_api::Type::Void,
+      NativeType::I64 => fast_api::Type::Int64,
+      NativeType::U64 => fast_api::Type::Uint64,
+      NativeType::ISize => fast_api::Type::Int64,
+      NativeType::USize | NativeType::Pointer | NativeType::Function => {
+        fast_api::Type::Uint64
```

Comment:
I want to double-check: this is (or would be if they existed) incorrect for 32 bits builds?

Reply:
Yes